src/linear_solver/NNLS_solver.cpp

```diff
@@ -0,0 +1,364 @@
+    if (P[i] == 1) {
+      P_mat.InsertGlobalValues(GlobalRow, 1, &posOne , &i);
+    }
+  }
+}
+
+void NNLS_solver::PositiveSetMatrix(Epetra_CrsMatrix &P_mat){
+  // Create matrix P_mat which contains the positive set of columns in A
+
+  // Create map between index_set and the columns to be added to P_mat
+  std::vector<int> colMap(A_.NumGlobalCols());
+  int numCol = 0;
+  for(int j = 0; j < A_.NumGlobalCols(); j++){
+    int idx = index_set[j];
+    if (P[idx] == 1) {
+      colMap[idx] = numCol;
+      numCol++;
+    }
+  }
+
+  // Fill Epetra_CrsMatrix P_mat with columns of A in set P
+  for(int i =0; i < A_.NumGlobalRows(); i++){
+    double *row = new double[A_.NumGlobalCols()];
+    int numE;
+    const int globalRow = A_.GRID(i);
+    A_.ExtractGlobalRowCopy(globalRow, A_.NumGlobalCols(), numE , row);
+    for(int j = 0; j < A_.NumGlobalCols(); j++){
+      if (P[j] == 1) {
+        P_mat.InsertGlobalValues(i, 1, &row[j] , &colMap[j]);
+        
+      }
+    }
+  }
+}
+
+void NNLS_solver::SubIntoX(Epetra_Vector &temp){
+  // Substitute new values into the solution vector
+  std::vector<int> colMap(A_.NumGlobalCols());
+  int numCol = 0;
+  for(int j = 0; j < x_.GlobalLength(); j++){
+    int idx = index_set[j];
+    if (P[idx] == 1) {
+      colMap[idx] = numCol;
+      numCol++;
+    }
+  }
+  for(int j = 0; j < x_.GlobalLength(); j++){
+    if (P[j] == 1) {
+      x_[j] = temp[colMap[j]];
+    }
+  }
+}
+
+void NNLS_solver::AddIntoX(Epetra_Vector &temp, double alpha){
+  // Add vector temp time scalar alpha into the vector x
+  std::vector<int> colMap(A_.NumGlobalCols());
+  int numCol = 0;
+  for(int j = 0; j < x_.GlobalLength(); j++){
+    int idx = index_set[j];
+    if (P[idx] == 1) {
+      colMap[idx] = numCol;
+      numCol++;
+    }
+  }
+  for(int j = 0; j < x_.GlobalLength(); j++){
+    if (P[j] == 1) {
+      x_[j] += alpha*(temp[colMap[j]] -x_[j]);
+    }
+  }
+}
+
+void NNLS_solver::moveToActiveSet(int idx){
+  // Move index at idx into the Active Set (Z set)
+  P[index_set(idx)] = 0;
+  Z[index_set(idx)] = 1; 
+
+  std::swap(index_set(idx), index_set(numInactive_ - 1));
+  numInactive_--;
+}
+
+void NNLS_solver::moveToInactiveSet(int idx){
+  // Move index at idx into the Inactive Set (P set)
+  P[index_set(idx)] = 1;
+  Z[index_set(idx)] = 0;
+
+  std::swap(index_set(idx), index_set(numInactive_));
+  numInactive_++;
+}
+
+bool NNLS_solver::solve(){
+  iter_ = 0;
+  numInactive_ = 0;
+
+  // Pre-mult by A^T
+  Epetra_CrsMatrix AtA(Epetra_DataAccess::View, A_.ColMap(), A_.NumMyCols());
+  EpetraExt::MatrixMatrix::Multiply(A_, true, A_, false, AtA);
+
+  Epetra_Vector Atb (A_.ColMap());
+  A_.Multiply(true, b_, Atb);
+
+  Epetra_Vector AtAx (A_.ColMap());
+  Epetra_Vector Ax (A_.RowMap());
+  Epetra_MultiVector gradient (A_.ColMap(), 1);
+  Epetra_MultiVector residual (A_.RowMap(), 1);
+
+  Eigen::VectorXd grad_eig (gradient.GlobalLength());
+  Epetra_Vector grad_col (A_.ColMap());
+
+  // OUTER LOOP
+  while(true){
```

Comment:
Would it be possible to use exit flags for the outer and inner loop? Having clear exit conditions might help future users understand the code easier. If exit flags aren't a viable option for the actions being performed here, then there's no need to change it! 😄

Reply:
Modified the inner loop to use a flag, but left the outer the same due to the use of multiple return statements